paimon-core/src/main/java/org/apache/paimon/io/FileIndexWriter.java

```diff
@@ -191,4 +237,76 @@ public byte[] serializedBytes() {
             return fileIndexWriter.serializedBytes();
         }
     }
+
+    /** File index writer for map data type. */
+    private static class MapFileIndexMaintainer {
+
+        private final String columnName;
+        private final String indexType;
+        private final Options options;
+        private final DataType valueType;
+        private final Map<String, org.apache.paimon.fileindex.FileIndexWriter> indexWritersMap;
+        private final InternalArray.ElementGetter valueElementGetter;
+        private final int position;
+
+        public MapFileIndexMaintainer(
+                String columnName,
+                String indexType,
+                DataType keyType,
+                DataType valueType,
+                Options options,
+                int position) {
+            this.columnName = columnName;
+            this.indexType = indexType;
+            this.valueType = valueType;
+            this.options = options;
+            this.position = position;
+            this.indexWritersMap = new HashMap<>();
+            this.valueElementGetter = InternalArray.createElementGetter(valueType);
+
+            DataTypeRoot rootType = keyType.getTypeRoot();
+            if (rootType != DataTypeRoot.CHAR && rootType != DataTypeRoot.VARCHAR) {
+                throw new IllegalArgumentException(
+                        "Only support map data type with key field of CHAR、VARCHAR、STRING.");
+            }
+        }
+
+        public void write(InternalRow row) {
+            InternalMap internalMap = row.getMap(position);
+            InternalArray keyArray = internalMap.keyArray();
+            InternalArray valueArray = internalMap.valueArray();
+
+            for (int i = 0; i < keyArray.size(); i++) {
+                String key = keyArray.getString(i).toString();
+                org.apache.paimon.fileindex.FileIndexWriter writer =
+                        indexWritersMap.getOrDefault(key, null);
+                if (writer != null) {
+                    writer.write(valueElementGetter.getElementOrNull(valueArray, i));
+                }
+            }
+        }
+
+        public void add(String nestedKey, Options nestedOptions) {
+            indexWritersMap.put(
+                    nestedKey,
+                    FileIndexer.create(
+                                    indexType,
+                                    valueType,
+                                    new Options(options.toMap(), nestedOptions.toMap()))
+                            .createWriter());
+        }
+
+        public String getIndexType() {
+            return indexType;
+        }
+
+        public Map<String, byte[]> serializedBytes() {
+            Map<String, byte[]> result = new HashMap<>();
+            indexWritersMap.forEach(
+                    (k, v) ->
+                            result.put(
+                                    FileIndexCommon.toMapKey(columnName, k), v.serializedBytes()));
```

Comment:
You need to consider no value in this file. We should have a special mechanism to compress this situation.